### PR TITLE
Merging states

### PR DESCRIPTION
**Pull request by ThomasWerthenbach**
loops can now be merged in two ways, namely singular (subsequent) states or recurring states (view python doc)
It is also possible to merge two arbitrary nodes

### CODE REVIEW

File: whatthelog/prefixtree/graph.py [hunk 1]
```diff
@@ -45,6 +49,144 @@ def get_state_by_hash(self, state_hash: int):
...
+        """
+        assert self.start_node is not None
+        stack: List[Tuple[State, List]] = [(self.start_node, [])]
+        visited = set()
+
+        current_unique = []
+
+        while len(stack) != 0:
+            state, unique = stack.pop()
+            if len(unique) is not 0:
```
Comment (gcalin) [inline whatthelog/prefixtree/graph.py]:
```suggestion
            if len(unique) != 0:
```

The is operator should only be used for objects with the same reference.

File: whatthelog/prefixtree/graph.py [hunk 2]
```diff
@@ -45,6 +49,134 @@ def get_state_by_hash(self, state_hash: int):
...
+            if state in current_unique:  # Should not occur, but does not hurt.
+                continue
+            else:
+                outgoing = [x for x in self.get_outgoing_states(state) if x not in current_unique]
+
+                arr = list(filter(lambda x: state.is_equivalent(x), current_unique))
+
+                if len(arr) > 0:
+                    self.merge_states(arr[0], state)
+                    state = arr[0]
```
Comment (ThomasWerthenbach) [inline whatthelog/prefixtree/graph.py]:
```suggestion
```

File: whatthelog/prefixtree/graph.py [hunk 3]
```diff
@@ -45,6 +49,158 @@ def get_state_by_hash(self, state_hash: int):
...
+
+                # Get all unvisited neighbours
+                outgoing = [x for x in self.get_outgoing_states(state) if x not in current_unique]
+
+                # Get all equivalent states from the unique list
+                arr = list(filter(lambda x: state.is_equivalent(x), current_unique))
+
+                if len(arr) > 0:
+                    # If any are equivalent, merge the first one?
+                    self.merge_states(arr[0], state)
```
Comment (gcalin) [inline whatthelog/prefixtree/graph.py]: Why does this only merge the first of the equivalent states?

Comment (ThomasWerthenbach) [reply whatthelog/prefixtree/graph.py]: Because there should only be one unique state, if any.

**Pull request merged**